### DISCUSSION TITLE
Optional Body, params, or querystring

### DISCUSSION THREAD

**Discussion by Armania** [fastify/fastify#5202] [Q&A] [answered]
Hi Fastify Community,

Question, if your provide a schema for the Body, Params, or Query, does your request have to provide data for these properties

Example I want my api route to allow providing an "id" via the Body or a querystring. But when I provide a schema for both, querystring and body, and I pass an id value id via one and not the other I get an error.

So, if I send an id via DELETE /?id=123 , I get an error. ~body does not match schema.

Do I have to send a body with every request bc I proivded a Body schema via the route option?

**Comment by Armania** [comment]
{
    "statusCode": 400,
    "code": "FST_ERR_VALIDATION",
    "error": "Bad Request",
    "message": "body must be object"
}

**Comment by mcollina** [comment]
I would try with an oneOf or anyOf    Operators, adding `null`. I'm not sure it would work. Overall Fastify does not offer a way to support this case (a value in either query or body).

**Reply by Armania** [reply]
Thank you for the quick response!

I have been trying anyOf , to no avail,  but not oneOf, will give that a try and see. ty!

**Comment by gurgunday** [comment]
From the error maybe you’re sending an empty (null) body or something, and not sure if that’s possible but you might also try to add nullable true to your body definition

**Reply by Armania** [reply]
Actually, it works if I send an empty object "{}" but not if at all.

**Reply by gurgunday** [reply]
Interesting... it shouldn't do that

You sure you haven't made the body a required property somehow?

Can you pass me the validation schema?

**Reply by Armania** [reply]
```javascript
const body = {
  anyOf: [
    {
      title: "users",
      type: "object",
      description: "list of users",
      properties: {
        userid: User.properties.userid // { type: "string" }
      },
    },
    {
      type: "null",
    },
  ],
};
```

**Reply by gurgunday** [reply, marked as answer]
There you go this fixes it:

```js
const body = {
  title: "users",
  type: "object",
  nullable: true, // add nullable
  description: "list of users",
  properties: {},
};
```

**Reply by gurgunday** [reply]
The reason is in `lib/validation.js` line 109 — we pass null if a validated property (body) is undefined

**Reply by Armania** [reply]
That did the trick, 🥳 

Also, thank you for the explanation.
